lib/experimental/style-variations.php

```diff
@@ -0,0 +1,185 @@
+					return current_user_can( 'edit_posts' );
+				},
+			)
+		);
+	}
+}
+add_action( 'init', 'gutenberg_register_style_variation_meta' );
+
+/**
+ * Get a style variation by ID.
+ *
+ * @param int $style_variation_id The style variation post ID.
+ * @return array|null The style variation data or null if not found.
+ */
+function gutenberg_get_style_variation( $style_variation_id ) {
+	$post = get_post( $style_variation_id );
+
+	if ( ! $post || 'wp_global_styles' !== $post->post_type ) {
+		return null;
+	}
+
+	// Verify it's a style variation.
+	$is_style_variation = get_post_meta( $style_variation_id, GUTENBERG_STYLE_VARIATION_META_KEY, true );
+	if ( ! $is_style_variation ) {
+		return null;
+	}
+
+	return array(
+		'id'      => $post->ID,
+		'title'   => $post->post_title,
+		'content' => json_decode( $post->post_content, true ),
+	);
+}
+
+/**
+ * Get the connected style variation ID for a post.
+ *
+ * @param int $post_id The post ID.
+ * @return int The connected style variation ID, or 0 if none.
+ */
+function gutenberg_get_connected_style_variation_id( $post_id ) {
+	return (int) get_post_meta( $post_id, GUTENBERG_CONNECTED_STYLE_VARIATION_META_KEY, true );
+}
+
+/**
+ * Get a WP_Theme_JSON_Gutenberg object from a style variation's content.
+ *
+ * @param array $style_variation The style variation data.
+ * @return WP_Theme_JSON_Gutenberg The theme JSON object.
+ */
+function gutenberg_get_style_variation_theme_json( $style_variation ) {
+	$content = $style_variation['content'];
+
+	// Ensure content is valid.
+	if ( ! is_array( $content ) ) {
+		$content = array();
+	}
+
+	// Remove the flag that marks it as user theme JSON.
+	unset( $content['isGlobalStylesUserThemeJSON'] );
+
+	return new WP_Theme_JSON_Gutenberg( $content, 'custom' );
+}
+
+/**
+ * Add style variation data to the block editor settings.
+ *
+ * This provides the editor with information about connected style variations
+ * so it can load the correct styles via the JavaScript side.
+ *
+ * @param array $settings The block editor settings.
+ * @return array Modified settings.
+ */
+function gutenberg_add_style_variations_to_editor_settings( $settings ) {
+	$post_id = 0;
+
+	// Get the post ID from the editor context.
+	if ( isset( $settings['postId'] ) ) {
+		$post_id = $settings['postId'];
+	}
+
+	if ( ! $post_id ) {
+		return $settings;
+	}
+
+	// Get the connected style variation for this post.
+	$connected_style_variation_id = gutenberg_get_connected_style_variation_id( $post_id );
+
+	$settings['connectedStyleVariationId'] = $connected_style_variation_id;
+
+	return $settings;
+}
+add_filter( 'block_editor_settings_all', 'gutenberg_add_style_variations_to_editor_settings' );
+
+/**
+ * Enqueue style variation CSS on the frontend.
+ *
+ * This ensures connected style variations are properly applied on singular pages.
+ */
+function gutenberg_enqueue_style_variation_css() {
+	// Only run on singular pages.
+	if ( ! is_singular() ) {
```

Comment:
Similar to [below](https://github.com/WordPress/gutenberg/pull/73978/files#r2627080244)

I was wondering if style variations would apply site wide one day. E.g, a global styles post could be promoted from a style variation to become the main global styles post. Or is the per-post-only model the intended long-term design?
 
I think @youknowriad's issue is relevant.

- https://github.com/WordPress/gutenberg/issues/62686

If yes to  alternative "style variations" for themes, then this could blur the distinction between `root/globalStyles` and `postType/wp_global_styles` in the entity system.

Currently, both use the same REST endpoint (`/wp/v2/global-styles/{id}`) but are tracked as different entities

Once again, I don't have the full picture here.